File: dinosub.py
import os
import requests
from concurrent.futures import ThreadPoolExecutor as Pool

def install_required_modules():
    [__import__(m) for m in ["requests"] if not __import__(m, globals(), locals(), [], 0)]

def clear_terminal():
    os.system("cls" if os.name == "nt" else "clear")

def get_subdomains(domain):
    response = requests.get(f"https://crt.sh/?q=%.{domain}&output=json")
    return {entry['name_value'] for entry in response.json()} if response.status_code == 200 else set()

def save_subdomains_to_file(subdomains, filename):
    folder_path = os.path.join(os.path.dirname(__file__), "subdomains")
    os.makedirs(folder_path, exist_ok=True)
    file_path = os.path.join(folder_path, filename)
    with open(file_path, 'w') as f:
        f.write('\n'.join(subdomains))
    return os.path.abspath(file_path)

def main():
    install_required_modules()
    clear_terminal()
    print("DinoSub (Python) v1.2.1\nAuthor: https://github.com/daniisaahir\n")
    domain = input("Target Domain: ")
    subdomains = Pool().submit(get_subdomains, domain).result()
    print(f"Subdomains found: {len(subdomains)}\n{'-' * 36}")
    print(*subdomains, sep='\n')
    if input("Save results to txt file? (y/n): ").lower() == 'y':
        filename = input("Filename: ")
        if not filename.endswith(".txt"):
            filename += ".txt"
        saved_file_path = save_subdomains_to_file(subdomains, filename)
        print(f"Saved to: {saved_file_path}")

if __name__ == "__main__":
    main()

File: dinosub.js
const fs = require('fs');
const readline = require('readline');
const { promisify } = require('util');
const { exec } = require('child_process');
const fetch = require('isomorphic-fetch');
const path = require('path');

const executeCommand = promisify(exec);

function clearTerminal() {
  process.stdout.write('\x1Bc');
}

function prompt(question) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer);
    });
  });
}

async function getSubdomains(domain) {
  const response = await fetch(`https://crt.sh/?q=%.${domain}&output=json`);
  if (response.ok) {
    const data = await response.json();
    return new Set(data.map((entry) => entry.name_value));
  }
  return new Set();
}

function saveSubdomainsToFile(subdomains, filename) {
  const folderPath = path.join(__dirname, 'subdomains');
  fs.mkdirSync(folderPath, { recursive: true });
  const filePath = path.join(folderPath, filename);
  fs.writeFileSync(filePath, [...subdomains].join('\n'));
  return path.resolve(filePath);
}

async function main() {
  clearTerminal();
  console.log('DinoSub (Node.js) v1.2.1\nAuthor: https://github.com/daniisaahir\n');

  const domain = await prompt('Target Domain: ');
  const subdomains = await getSubdomains(domain);
  console.log(`Subdomains found: ${subdomains.size}\n${'-'.repeat(36)}`);
  console.log([...subdomains].join('\n'));

  const saveResults = (await prompt('Save results to txt file? (y/n): ')).toLowerCase();
  if (saveResults === 'y') {
    const filename = await prompt('Filename: ');
    const validFilename = filename.endsWith('.txt') ? filename : `${filename}.txt`;
    const savedFilePath = saveSubdomainsToFile(subdomains, validFilename);
    console.log(`Saved to: ${savedFilePath}`);
  }
}

main()
  .catch((error) => {
    console.error('An error occurred:', error);
    process.exit(1);
  });
